test_plus/test.py

```diff
@@ -381,6 +383,24 @@ def __init__(self, *args, **kwargs):
         self.client_class = get_api_client()
         super(APITestCase, self).__init__(*args, **kwargs)
 
+    # APITest requests now default to JSON
+    def request(self, method, url_name, *args, **kwargs):
+        data = kwargs.get("data")
+        kwargs["data"] = json.dumps(data) if data is not None else None
+        return getattr(self, method)(url_name, *args, **kwargs)
+
+    def post(self, url_name, *args, **kwargs):
+        return self.request("post", url_name, *args, **kwargs)
+
+    def get(self, url_name, *args, **kwargs):
+        return self.request("get", url_name, *args, **kwargs)
+
+    def put(self, url_name, *args, **kwargs):
+        return self.request("put", url_name, *args, **kwargs)
+
+    def delete(self, url_name, *args, **kwargs):
+        return self.request("delete", url_name, *args, **kwargs)
+
```

Comment:
I think you are going to need something like this instead:

```suggestion
    def request(self, method, url_name, *args, **kwargs):
        data = kwargs.get("data")
        kwargs["data"] = json.dumps(data) if data is not None else None

        extra = kwargs.get("extra")
        if extra and "format" not in extra:
            kwargs["extra"].update({"content_type": "application/json"})

        return super().request(method, url_name, *args, **kwargs)
```

The HTTP verbs will still pass through but the request() will default to JSON. I added the lookup format because it was erroring for me if I didn't locally.

Reply:
@frankwiles @epicserve do either of you have suggestions around this? I saw the comments/suggestions on #81 and I tried that but ran into other issues when I added a test. 

```
diff --git a/test_project/test_app/tests/test_unittests.py b/test_project/test_app/tests/test_unittests.py
index c467d6d..60298cc 100644
--- a/test_project/test_app/tests/test_unittests.py
+++ b/test_project/test_app/tests/test_unittests.py
@@ -628,7 +628,14 @@ class TestAPITestCaseDRFInstalled(APITestCase):
 
     def test_post(self):
         data = {'testing': {'prop': 'value'}}
-        self.post('view-json', data=data, extra={'format': 'json'})
+        response = self.post('view-json', data=data)
+        assert response.headers["Content-Type"] == "application/json"
+        self.response_200()
+
+    def test_post_with_format(self):
+        data = {'testing': {'prop': 'value'}}
+        response = self.post('view-json', data=data, extra={'format': 'json'})
+        assert response.headers["Content-Type"] == "application/json"
         self.response_200()
```

If possible, I'd prefer a solution that doesn't require maintaining a list of http verbs since that list was already missing a few. 

I also ran into some errors because tests blow up if we have both `format` and `content_type="application/json"` passed in. So it's possible my suggestion should do a better job of making it work for whatever variable vs. always dumping `format`. 

Thoughts?

Suggestions?

Reply:
@jefftriplett,

I get what you're saying about not wanting duplication for methods with the same REST verb (e.g. `self.post()` vs `self.post_json()`). I could go either way, honestly, but I think if you were to use `self.post()` to do both JSON/AJAX type requests and you want the default to be JSON/AJAX, then you'll need to make a decision on if it's worth it to make a breaking change and create a new major release.

For me personally, I like my IDE to show the two choices and it makes it a nicer DX to be able to choose between `self.post()` and `self.post_json()` in my IDE.

If you do go with just one method then I think I would want to see the methods be more explicit, which would also improve the DX.

Example:
```python
def post(self, url_name_or_url: str, /, *, data: Optional[dict] = None, format: str = 'JSON', extra: Optional[dict] = None, follow: bool = False):
    data = data | {}
    extra = data | {}
    return self.request('post', url_name, data=data, format=format, extra=extra, follow=follow)
```

cc: @frankwiles

Reply:
@epicserve I think the IDE suggestion is worth doing as a feature/follow-up. I'm not really sure the best way to added typed support (thinking about the older Python versions we support) but I'm +1

We are making changes to DRF's `rest_framework.test.APIClient` vs. Django's TestCase, so in theory it should already be json/ajax friendly. Do you think it's common that people use `tp_api` as a fixture instead of `tp` who would try to use it for non-JSON calls? 

If it's really measurable we could issue a deprecation warning and bump a few versions to get there.

Reply:
@jefftriplett, we don't use fixtures at all. We use Pytest + Class-based tests, which extend our base test, which extends `test_plus.TestCase`. I see now that you're extending APITestCase, which wasn't obvious initially because the diff didn't show the class. It seems like your solution shouldn't break anything since we're not extending from APITestCase. It does make me think that we should probably consider extending from a separate class when working on endpoint tests.

Reply:
Thank you for the feedback. I suspect this will be relatively painless then. I reworked the logic for `format` and `content_type` so it should respect what is passed in. The only usecase we aren't dealing with is allowing you to force the error when you provide both. I'll noodle on that some more.